File: examples/CBuilder/Compound/Building/Mainform.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Mainform.h"
#include "datamod.h"
#include "BuildingClasses.hpp"
//#include "BoldGUI,
#include "PersonAutoFormUnit.h"
#include <stdlib.h>
#include <Dialogs.hpp>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "BoldAbstractModel"
#pragma link "BoldModel"
#pragma link "BoldSubscription"
#pragma link "BoldEdit"
#pragma link "BoldAbstractListHandle"
#pragma link "BoldAFPPluggable"
#pragma link "BoldCursorHandle"
#pragma link "BoldGrid"
#pragma link "BoldHandles"
#pragma link "BoldListHandle"
#pragma link "BoldRootedHandles"
#pragma link "BoldCheckBox"
#pragma link "BoldExpressionHandle"
#pragma link "BoldListBox"
#pragma link "BoldControlPack"
#pragma link "BoldElements"
#pragma link "BoldStringControlPack"
#pragma link "BoldActions"
#pragma link "BoldDBActions"
#pragma link "BoldHandleAction"
#pragma link "BoldPersistenceNotifier"
#pragma link "BoldIBDatabaseAction"
#pragma link "BoldUndoInterfaces"
#pragma resource "*.dfm"
Tallform *allform;
//---------------------------------------------------------------------------
__fastcall Tallform::Tallform(TComponent* Owner)
        : TForm(Owner)
{
 // Currentpopupcomponent = new TComponent(this);
}
//---------------------------------------------------------------------------
void __fastcall Tallform::PopupPopup(TObject *Sender)
{
  if (dynamic_cast <TMenuItem*> (Sender))
  {
   Currentpopupcomponent = ActiveControl;
  }
  else if(dynamic_cast <TPopupMenu*> (Sender))
  {
   Currentpopupcomponent = ((TPopupMenu*)Sender)->PopupComponent;
  }
  else
  {
   Currentpopupcomponent = NULL;
  }
}
//---------------------------------------------------------------------------
void __fastcall Tallform::NewPersonClick(TObject *Sender)
{
  TPerson *nyperson = new TPerson(DataModule1->BoldSystemHandle1->System);
}
//---------------------------------------------------------------------------
void __fastcall Tallform::DeleteCurrentObjectClick(TObject *Sender)
{
  TBoldObject *BoldObject;
  TBoldListHandle *ListHandle;
  TBoldListBox *ListBox;
  TBoldGrid *Grid;

  if ((ListBox = dynamic_cast <TBoldListBox*> (Currentpopupcomponent)) != 0)
  {
    ListHandle = (TBoldListHandle*)ListBox->BoldHandle;
    if (ListHandle)
      BoldObject = ListBox->BoldHandle->CurrentBoldObject;
  }
  else if ((Grid = dynamic_cast <TBoldGrid*> (Currentpopupcomponent)) != 0)
  {
    ListHandle = (TBoldListHandle*)Grid->BoldHandle;
    if (ListHandle)
      BoldObject = Grid->BoldHandle->CurrentBoldObject;
  }
  else
    BoldObject = NULL;

  if (BoldObject)
    BoldObject->Delete();
}
//---------------------------------------------------------------------------
void __fastcall Tallform::ShowInOwnWindow(TObject *Sender)
{
  TBoldObject *BoldObject;
  TBoldListHandle *ListHandle;
  TBoldListBox *ListBox;
  TBoldGrid *Grid;

  if ((ListBox = dynamic_cast <TBoldListBox*> (Currentpopupcomponent)) != 0)
  {
    ListHandle = (TBoldListHandle*)ListBox->BoldHandle;
    if (ListHandle)
      BoldObject = ListBox->BoldHandle->CurrentBoldObject;
  }
  else if ((Grid = dynamic_cast <TBoldGrid*> (Currentpopupcomponent)) != 0)
  {
    ListHandle = (TBoldListHandle*)Grid->BoldHandle;
    if (ListHandle)
      BoldObject = Grid->BoldHandle->CurrentBoldObject;
  }
  else
    BoldObject = NULL;

  if (BoldObject)
    AutoFormProviderRegistry()->FormForElement(BoldObject)->Show();
}
//---------------------------------------------------------------------------
void __fastcall Tallform::CheckBox1Click(TObject *Sender)
{
  if (CheckBox1->Checked)
    blhAllPerson->BoldFilter = DataModule1->IsRichFilter;
  else
    blhAllPerson->BoldFilter = NULL;
}
//---------------------------------------------------------------------------
void __fastcall Tallform::CheckBox2Click(TObject *Sender)
{
  if (CheckBox2->Checked)
    blhAllPerson->BoldComparer = DataModule1->NameComparer;
  else
    blhAllPerson->BoldComparer = NULL;
}
//---------------------------------------------------------------------------
void __fastcall Tallform::FormCloseQuery(TObject *Sender, bool &CanClose)
{
  int status;
  if (!DataModule1->BoldSystemHandle1->Active)
    exit(status);
  if (DataModule1->BoldSystemHandle1->System->BoldDirty)
  {
    if (MessageDlg("You have unsaved changes, quit anyway?",
                  mtConfirmation, TMsgDlgButtons() << mbYes << mbNo, 0) == mrYes)
      DataModule1->BoldSystemHandle1->System->Discard();
    else
      CanClose = false;
  }
}
//---------------------------------------------------------------------------
void __fastcall Tallform::SingleItemRemove(TObject *Sender)
{
  TBoldEdit *ec;
  if ((ec = dynamic_cast <TBoldEdit*> (Currentpopupcomponent)) != 0)
  {
    if (dynamic_cast <TBoldObjectReference*> (ec->BoldHandle->Value))
      ((TBoldObjectReference*)ec->BoldHandle->Value)->BoldObject = NULL;
  }
}
//---------------------------------------------------------------------------
void __fastcall Tallform::Remove(TObject *Sender)
{
  TBoldListBox *lb;
  if ((lb = dynamic_cast <TBoldListBox*> (Currentpopupcomponent)) != 0)
  {
    if (lb->BoldHandle)
      lb->BoldHandle->RemoveCurrentElement();
  }
}
//---------------------------------------------------------------------------
void __fastcall Tallform::newBuildingClick(TObject *Sender)
{
  CurrentBuildingHandle()->List->AddNew();
}
//---------------------------------------------------------------------------
TBoldListHandle* __fastcall Tallform::CurrentBuildingHandle()
{
  if (PageControl1->ActivePage == tabBuilding)
    return blhAllBuilding;
  else
    return blhAllResidentialBuilding;
}
//---------------------------------------------------------------------------

void __fastcall Tallform::HighRentRendererSetColor(TBoldElement *Element,
      TColor &AColor, int Representation, AnsiString Expression)
{
  TResidential_Building *building;
  if ((building = dynamic_cast <TResidential_Building*> (Element)) != 0)
  {
    if (building->TotalRent >= 1500)
      AColor = clSilver;
    else
      AColor = clWhite;
  }
}
//---------------------------------------------------------------------------

void __fastcall Tallform::HighRentRendererSetFont(TBoldElement *Element,
      TFont *AFont, int Representation, AnsiString Expression)
{
  TResidential_Building *building;
  if ((building = dynamic_cast <TResidential_Building*> (Element)) != 0)
  {
    if (building->TotalRent >= 1500)
      AFont->Color = clSilver;
    else
      AFont->Color = clWhite;
  }
}
//---------------------------------------------------------------------------

void __fastcall Tallform::bsrAddressSetColor(TBoldElement *Element,
      TColor &AColor, int Representation, AnsiString Expression)
{
  TBuilding *building;
  if (Element)
  {
    if ((building = dynamic_cast <TBuilding*> (Element)) != 0)
    {
      if (building->Address.Pos("Bold") > 0)
        AColor = clAqua;
    }
  }
}
//---------------------------------------------------------------------------

void __fastcall Tallform::bsrAddressSetFont(TBoldElement *Element,
      TFont *AFont, int Representation, AnsiString Expression)
{
  TBuilding *building;
  if (Element)
  {
    if ((building = dynamic_cast <TBuilding*> (Element)) != 0)
    {
      if (building->M_Address->AsString.Pos("Bold") > 0)
        AFont->Style = AFont->Style + TFontStyles()<< fsBold;//fsBold];
      if (building->M_Address->AsString.Pos("Rose") > 0)
        AFont->Color = clRed;
      if (building->M_Address->AsString.Pos("Select") > 0)
        AFont->Color = clGreen;
    }
  }
}
//---------------------------------------------------------------------------

AnsiString __fastcall Tallform::bsrResidentsTotalAssetsGetAsString(
      TBoldElement *Element, int Representation, AnsiString Expression)
{
  int i;
  Currency Sum = 0;
  TResidential_Building *building;

  if ((building = dynamic_cast <TResidential_Building*> (Element)) != 0)
  {
    for (i=0; i<building->Residents->Count; i++)
      Sum = Sum + building->Residents->BoldObjects[i]->Assets;
  }
  return CurrToStr(Sum);
}
//---------------------------------------------------------------------------

void __fastcall Tallform::bsrResidentsTotalAssetsSubscribe(
      TBoldElement *Element, int Representation, AnsiString Expression,
      TBoldSubscriber *Subscriber)
{
  int i;
  TResidential_Building *building;
  TBoldExternalVariableList* VariableList = NULL;

  if ((building = dynamic_cast <TResidential_Building*> (Element)) != 0)
  {
    building->SubscribeToExpression("residents", Subscriber, true, false, VariableList);
    for(i=0; i<building->Residents->Count; i++)
      building->Residents->BoldObjects[i]->SubscribeToExpression("assets", Subscriber, false, false, VariableList);
  }
}
//---------------------------------------------------------------------------

AnsiString __fastcall Tallform::bsrRentPerResidentGetAsString(
      TBoldElement *Element, int Representation, AnsiString Expression)
{
  AnsiString Result = "";

  TResidential_Building *building;
  if ((building = dynamic_cast <TResidential_Building*> (Element)) != 0)
  {
    if (building->M_TotalRent->IsNull)
      Result = "***";
    else if ((building->Residents == NULL) || (building->Residents->Count == 0))
      Result = "No Residents";
    else
      Result = CurrToStr(building->TotalRent/building->Residents->Count);
  }
  return Result;
}
//---------------------------------------------------------------------------

void __fastcall Tallform::bsrRentPerResidentHoldsChangedValue(
      TBoldElement *Element, int Representation, AnsiString Expression,
      TBoldSubscriber *Subscriber)
{
  TResidential_Building *building;
  if ((building = dynamic_cast <TResidential_Building*> (Element)) != 0)
    building->M_TotalRent->RegisterModifiedValueHolder(Subscriber);
}
//---------------------------------------------------------------------------

bool __fastcall Tallform::bsrRentPerResidentMayModify(
      TBoldElement *Element, int Representation, AnsiString Expression,
      TBoldSubscriber *Subscriber)
{
  bool Result = false;

  TResidential_Building *building;
  if ((building = dynamic_cast <TResidential_Building*> (Element)) != 0)
  {
    if (building->Residents)
      Result = building->Residents->Count > 0;
  }
  return Result;
}
//---------------------------------------------------------------------------

void __fastcall Tallform::bsrRentPerResidentReleaseChangedValue(
      TBoldElement *Element, int Representation, AnsiString Expression,
      TBoldSubscriber *Subscriber)
{
  TResidential_Building *building;
  if ((building = dynamic_cast <TResidential_Building*> (Element)) != 0)
    building->M_TotalRent->UnRegisterModifiedValueHolder(Subscriber);
}
//---------------------------------------------------------------------------

void __fastcall Tallform::bsrRentPerResidentSetAsString(
      TBoldElement *Element, AnsiString NewValue, int Representation,
      AnsiString Expression)
{
  AnsiString v;
  v = NewValue; // avoid name clash
  TResidential_Building *building;
  if ((building = dynamic_cast <TResidential_Building*> (Element)) != 0)
    building->TotalRent = StrToCurr(v) * building->Residents->Count;
}
//---------------------------------------------------------------------------

void __fastcall Tallform::bsrRentPerResidentSubscribe(
      TBoldElement *Element, int Representation, AnsiString Expression,
      TBoldSubscriber *Subscriber)
{
  TBoldExternalVariableList* VariableList = NULL;
  if (dynamic_cast <TResidential_Building*> (Element))
  {
    Element->SubscribeToExpression("totalRent", Subscriber, false, false, VariableList);
    Element->SubscribeToExpression("residents", Subscriber, false, false, VariableList);
  }
}
//---------------------------------------------------------------------------

bool __fastcall Tallform::bsrRentPerResidentValidateCharacter(
      TBoldElement *Element, AnsiString Value, int Representation,
      AnsiString Expression)
{
  if (Value[1] == '0' || Value[1] == '1' || Value[1] == '2' || Value[1] == '3' ||
      Value[1] == '4' || Value[1] == '5' || Value[1] == '6' || Value[1] == '7' ||
      Value[1] == '8' || Value[1] == '9' || Value[1] == '-' || Value[1] == '+' ||
      Value[1] == 'e' || Value[1] == 'E' || Value[1] == DecimalSeparator)
      return true;

  return false;
}
//---------------------------------------------------------------------------

bool __fastcall Tallform::bsrRentPerResidentValidateString(
      TBoldElement *Element, AnsiString Value, int Representation,
      AnsiString Expression)
{
  try
  {
    StrToCurr(Value);
    return true;
  } catch(Exception &e)
  {
    return false;
  }
}
//---------------------------------------------------------------------------

void __fastcall Tallform::BoldActivateSystemAction1SystemOpened(
      TObject *Sender)
{
    Randomize();
}
//---------------------------------------------------------------------------

void __fastcall Tallform::btnChargeRentClick(TObject *Sender)
{
  TResidential_Building *residens;
  if ((residens = dynamic_cast <TResidential_Building*>(blhAllResidentialBuilding->CurrentBoldObject)) != 0)
    residens->ChargeRent();
}
//---------------------------------------------------------------------------


void __fastcall Tallform::rgNameRepresentationClick(TObject *Sender)
{
  switch(rgNameRepresentation->ItemIndex)
  {
    case 0: blbBuildingOwners->BoldRowProperties->Expression = "firstName";
            break;
    case 1: blbBuildingOwners->BoldRowProperties->Expression = "lastName";
            break;
    case 2: blbBuildingOwners->BoldRowProperties->Expression = "firstName + ' ' + lastName";
            break;
  }
}
//---------------------------------------------------------------------------

void __fastcall Tallform::BoldPlaceableAFP1GetFormClass(
      TBoldElement *Element, TFormClass &Result)
{
  Result = __classid(TPersonAutoForm);
}
//---------------------------------------------------------------------------

void __fastcall Tallform::BoldPlaceableAFP1RetrieveHandle(TForm *Form,
      TBoldReferenceHandle *&Result)

{
  Result = NULL;
  TPersonAutoForm *PersonForm = dynamic_cast<TPersonAutoForm*>(Form);
  if (PersonForm != NULL)
  {
    Result = PersonForm->brhPerson;
  }
}
//---------------------------------------------------------------------------

void __fastcall Tallform::btnCheckpointClick(TObject *Sender)
{
  DataModule1->BoldSystemHandle1->System->UndoHandlerInterface->SetCheckPoint();
}
//---------------------------------------------------------------------------

void __fastcall Tallform::btnUnDoClick(TObject *Sender)
{
 DataModule1->BoldSystemHandle1->System->UndoHandlerInterface->UndoLatest();
}
//---------------------------------------------------------------------------

void __fastcall Tallform::btnRedoClick(TObject *Sender)
{
  DataModule1->BoldSystemHandle1->System->UndoHandlerInterface->RedoLatest();
}
//---------------------------------------------------------------------------

File: examples/Delphi/Compound/HTTPPMapper/Client/mainform.pas
unit mainform;

interface

uses
  SysUtils,
  Classes,
  Graphics,
  Controls,
  Forms,
  Dialogs,
  Menus,
  StdCtrls,
  ExtCtrls,
  ComCtrls,
  Grids,
  ActnList,
  BoldSubscription,
  BoldElements,
  BoldSystem,
  BoldHandles,
  BoldListHandle,
  BoldAbstractListHandle,
  BoldCursorHandle,
  BoldExpressionHandle,
  BoldRootedHandles,
  BoldPersistenceNotifier,
  BoldAFP,
  BoldAFPDefault,
  BoldControlPack,
  BoldStringControlPack,
  BoldCheckBox,
  BoldGrid,
  BoldEdit,
  BoldNavigator,
  BoldLabel,
  BoldListBox,
  BoldHandleAction,
  BoldActions,
  BoldDBActions,
  BoldAFPPluggable,
  BoldReferenceHandle
  ;

type
  Tallform = class(TForm)
    BuildingPopup: TPopupMenu;
    newBuilding: TMenuItem;
    Delete1: TMenuItem;
    PersonPopup: TPopupMenu;
    NewPerson: TMenuItem;
    DeletePerson: TMenuItem;
    PersonGroup: TGroupBox;
    SingleLinkPopup: TPopupMenu;
    MenuItem5: TMenuItem;
    Label2: TLabel;
    Label3: TLabel;
    Label1: TLabel;
    FirstName: TLabel;
    btnUpdateDB: TButton;
    bedFirstName: TBoldEdit;
    PersonCount: TBoldEdit;
    bedPersonHome: TBoldEdit;
    blbPersonOwnedBuildings: TBoldListBox;
    MultiLinkPopup: TPopupMenu;
    MenuItem1: TMenuItem;
    bedLastName: TBoldEdit;
    Label6: TLabel;
    blhAllPerson: TBoldListHandle;
    blhAllBuilding: TBoldListHandle;
    blhOwnedBuildingsHandle: TBoldListHandle;
    blhOwners: TBoldListHandle;
    behHome: TBoldExpressionHandle;
    Showinownwindow1: TMenuItem;
    Showinownwindow2: TMenuItem;
    ShowInOwnWindow3: TMenuItem;
    bsrRentPerResident: TBoldAsStringRenderer;
    bedAssets: TBoldEdit;
    Label7: TLabel;
    bcbRich: TBoldCheckBox;
    CheckBox1: TCheckBox;
    CheckBox2: TCheckBox;
    bsrResidentsTotalAssets: TBoldAsStringRenderer;
    bsrAddress: TBoldAsStringRenderer;
    bgrPerson: TBoldGrid;
    pbdbNotification: TProgressBar;
    PageControl1: TPageControl;
    tabBuilding: TTabSheet;
    tabResidentialBuilding: TTabSheet;
    bgrResidentialBuilding: TBoldGrid;
    BuildingGroup: TGroupBox;
    Owners: TLabel;
    Resi: TLabel;
    Label5: TLabel;
    bedAddress2: TBoldEdit;
    blbBuildingResidents: TBoldListBox;
    blbBuildingOwners2: TBoldListBox;
    btnChargeRent: TButton;
    bgrBuilding: TBoldGrid;
    GroupBox1: TGroupBox;
    Label4: TLabel;
    Label9: TLabel;
    bedAddress: TBoldEdit;
    blbBuildingOwners: TBoldListBox;
    blhResidents: TBoldListHandle;
    blhAllResidentialBuilding: TBoldListHandle;
    blhOwners2: TBoldListHandle;
    rgNameRepresentation: TRadioGroup;
    BoldPersistenceProgressNotifier1: TBoldPersistenceProgressNotifier;
    ActionList1: TActionList;
    BoldActivateSystemAction1: TBoldActivateSystemAction;
    Button1: TButton;
    BoldUpdateDBAction1: TBoldUpdateDBAction;
    HighRentRenderer: TBoldAsStringRenderer;
    BoldPlaceableAFP1: TBoldPlaceableAFP;
    procedure newBuildingClick(Sender: TObject);
    procedure DeleteCurrentObject(Sender: TObject);
    procedure NewPersonClick(Sender: TObject);
    procedure Remove(Sender: TObject);
    procedure SingleItemRemove(Sender: TObject);
    procedure PopupPopup(Sender: TObject);
    procedure FormCloseQuery(Sender: TObject; var CanClose: Boolean);
    procedure ShowInOwnWindow(Sender: TObject);
    function ReadOnlyRendererObserverMayModify(
      Element: TBoldElement; representation: Integer;
      Expression: String; Subscriber: TBoldSubscriber): Boolean;
    function bsrRentPerResidentGetAsString(Element: TBoldElement;
      representation: Integer; Expression: String): string;
    function RendPerResidentRendererObserverMayModify(
      Element: TBoldElement; representation: Integer;
      Expression: String; Subscriber: TBoldSubscriber): Boolean;
    procedure bsrRentPerResidentHoldsChangedValue(
      Element: TBoldElement; representation: Integer;
      Expression: String; Subscriber: TBoldSubscriber);
    procedure bsrRentPerResidentReleaseChangedValue(
      Element: TBoldElement; representation: Integer;
      Expression: String; Subscriber: TBoldSubscriber);
    procedure bsrRentPerResidentSubscribe(Element: TBoldElement;
      representation: Integer; Expression: String;
      Subscriber: TBoldSubscriber);
    procedure bsrRentPerResidentSetAsString(Element: TBoldElement;
      value: string; representation: Integer; Expression: String);
    procedure rgNameRepresentationClick(Sender: TObject);
    procedure btnChargeRentClick(Sender: TObject);
    procedure CheckBox1Click(Sender: TObject);
    procedure CheckBox2Click(Sender: TObject);
    function bsrRentPerResidentMayModify(element: TBoldElement;
      representation: Integer; Expression: String;
      Subscriber: TBoldSubscriber): Boolean;
    function bsrRentPerResidentValidateCharacter(
      element: TBoldElement; value: String; representation: Integer;
      Expression: String): Boolean;
    function bsrRentPerResidentValidateString(
      element: TBoldElement; value: String; representation: Integer;
      Expression: String): Boolean;
    function bsrResidentsTotalAssetsGetAsString(element: TBoldElement;
      representation: Integer; Expression: String): String;
    procedure bsrResidentsTotalAssetsSubscribe(element: TBoldElement;
      representation: Integer; Expression: String;
      Subscriber: TBoldSubscriber);
    procedure bsrAddressSetColor(element: TBoldElement;
      var aColor: TColor; representation: Integer; Expression: String);
    procedure bsrAddressSetFont(element: TBoldElement;
      aFont: TFont; representation: Integer; Expression: String);
    procedure NewBuilding1Click(Sender: TObject);
    procedure BoldActivateSystemAction1SystemOpened(Sender: TObject);
    procedure HighRentRendererSetColor(Element: TBoldElement;
      var AColor: TColor; Representation: Integer; Expression: String);
    procedure HighRentRendererSetFont(Element: TBoldElement; AFont: TFont;
      Representation: Integer; Expression: String);
   { function BoldPlaceableAFP1GetFormClass(
      Element: TBoldElement): TFormClass;
    function BoldPlaceableAFP1RetrieveHandle(
      Form: TForm): TBoldReferenceHandle;}


  private
    { Private declarations }
    function CurrentBuildingHandle: TBoldListHandle;
  public
    { Public declarations }
  end;

var
  allform: Tallform;

implementation

uses
  BuildingClasses,
  BoldGUI,
  datamod;

{$R *.DFM}

var
  Currentpopupcomponent: TComponent;

procedure Tallform.newBuildingClick(Sender: TObject);
begin
  CurrentBuildingHandle.List.AddNew;
end;

procedure Tallform.NewPersonClick(Sender: TObject);
begin
  TPerson.Create(nil);
end;

procedure Tallform.DeleteCurrentObject(Sender: TObject);
var
  BoldObject: TBoldObject;
begin
  if (CurrentPopupComponent is TBoldListBox) and
     Assigned(TBoldListBox(CurrentPopupComponent).BoldHandle) then
    BoldObject := TBoldListBox(CurrentPopupComponent).BoldHandle.CurrentBoldObject
  else if (CurrentPopupComponent is TBoldGrid) and
          Assigned(TBoldGrid(CurrentPopupComponent).BoldHandle) then
    BoldObject := TBoldGrid(CurrentPopupComponent).BoldHandle.CurrentBoldObject
  else
    BoldObject := nil;

  if assigned(BoldObject) then
    BoldObject.Delete;
end;

procedure Tallform.Remove(Sender: TObject);
begin
  with Currentpopupcomponent as TBoldListBox do
    if Assigned(BoldHandle) then
      BoldHandle.RemoveCurrentElement;
end;

procedure Tallform.SingleItemRemove(Sender: TObject);
begin
  with Currentpopupcomponent as TBoldEdit do
   if BoldHandle.Value is TBoldObjectReference then
     TBoldObjectReference(BoldHandle.Value).BoldObject := nil;
end;

procedure Tallform.PopupPopup(Sender: TObject);
begin
   if (sender is TMenuItem) then
     Currentpopupcomponent  := ActiveControl
   else if sender is TPopupMenu then
     Currentpopupcomponent := TPopupMenu(sender).popupcomponent
   else
     Currentpopupcomponent := nil;
end;

procedure Tallform.FormCloseQuery(Sender: TObject; var CanClose: Boolean);
begin
  if not datamodule1.BoldSystemHandle1.Active then
    Exit;
  if (datamodule1.BoldSystemHandle1.System.BoldDirty) then
    if Messagedlg('You have unsaved changes, quit anyway?',
                  mtConfirmation, [mbYes, mbNo], 0) = mrYes then
      datamodule1.BoldSystemHandle1.System.Discard
    else
      Canclose := False;
end;

procedure Tallform.ShowInOwnWindow(Sender: TObject);
var
  BoldObject: TBoldObject;
begin
   if (CurrentPopupComponent is TBoldListBox) and
    Assigned(TBoldListBox(CurrentPopupComponent).BoldHandle) then
    BoldObject := TBoldListBox(CurrentPopupComponent).BoldHandle.CurrentBoldObject
  else  if (CurrentPopupComponent is TBoldGrid) and
    Assigned(TBoldGrid(CurrentPopupComponent).BoldHandle) then
    BoldObject := TBoldGrid(CurrentPopupComponent).BoldHandle.CurrentBoldObject
  else
    BoldObject := nil;

  if assigned(BoldObject) then
    AutoFormProviderRegistry.FormForElement(BoldObject).Show;
end;

function Tallform.ReadOnlyRendererObserverMayModify(
  Element: TBoldElement; representation: Integer; Expression: String;
  Subscriber: TBoldSubscriber): Boolean;
begin
  Result := True;
end;

function Tallform.bsrRentPerResidentGetAsString(
  Element: TBoldElement; representation: Integer;
  Expression: String): string;
begin
  result := '';
  if element is TResidential_Building then
    with Element as TResidential_Building do
      if M_TotalRent.IsNull then
        Result := '***'
      else if Residents.Count = 0 then
        Result := 'No Residents'
      else
         Result := CurrToStr(TotalRent/Residents.Count);
end;

function Tallform.RendPerResidentRendererObserverMayModify(
  Element: TBoldElement; representation: Integer; Expression: String;
  Subscriber: TBoldSubscriber): Boolean;
begin
  result := false;
  if element is TResidential_Building then
    with Element as TResidential_Building do
      Result := Residents.Count <> 0;
end;

procedure Tallform.bsrRentPerResidentHoldsChangedValue(
  Element: TBoldElement; representation: Integer; Expression: String;
  subscriber: TBoldsubscriber);
begin
  if element is TResidential_Building then
    with Element as TResidential_Building do
      M_TotalRent.RegisterModifiedValueHolder(subscriber);
end;

procedure Tallform.bsrRentPerResidentReleaseChangedValue(
  Element: TBoldElement; representation: Integer; Expression: String;
  subscriber: TBoldSubscriber);
begin
  if element is TResidential_Building then
    with Element as TResidential_Building do
      M_TotalRent.UnRegisterModifiedValueHolder(Subscriber);
end;

procedure Tallform.bsrRentPerResidentSubscribe(
  Element: TBoldElement; representation: Integer; Expression: String;
  Subscriber: TBoldSubscriber);
begin
  if element is TResidential_Building then
    with Element do
    begin
      SubscribeToExpression('totalRent', Subscriber, False);
      SubscribeToExpression('residents', Subscriber, False);
    end;
end;

procedure Tallform.bsrRentPerResidentSetAsString(
  Element: TBoldElement; value: string; representation: Integer;
  Expression: String);
var
  v: string;
begin
  v := value; {avoid name clash}
  if element is TResidential_Building then
    with Element as TResidential_Building do
      TotalRent := StrToCurr(v) * Residents.Count;
end;

procedure Tallform.rgNameRepresentationClick(Sender: TObject);
begin
  with blbBuildingOwners.BoldRowProperties do
    case rgNameRepresentation.ItemIndex of
      0: Expression  := 'firstName';
      1: Expression  := 'lastName';
      2: Expression  := 'firstName + '' '' + lastName';
    end;
end;

procedure Tallform.btnChargeRentClick(Sender: TObject);
begin
  if blhAllResidentialBuilding.CurrentBoldObject is TResidential_Building then
    (blhAllResidentialBuilding.CurrentBoldObject as TResidential_Building).ChargeRent;
end;

procedure Tallform.CheckBox1Click(Sender: TObject);
begin
  if CheckBox1.Checked then
    blhAllPerson.BoldFilter := datamodule1.IsRichFilter
  else
    blhAllPerson.BoldFilter := nil;
end;

procedure Tallform.CheckBox2Click(Sender: TObject);
begin
   if CheckBox2.Checked then
    blhAllPerson.BoldComparer := datamodule1.NameComparer
  else
    blhAllPerson.BoldComparer := nil;
end;

function Tallform.bsrRentPerResidentMayModify(
  element: TBoldElement; representation: Integer;
  Expression: String; Subscriber: TBoldSubscriber): Boolean;
begin
  Result := False;
  if element is TResidential_Building then
    with element as TResidential_Building do
      Result := Residents.Count > 0;
end;

function Tallform.bsrRentPerResidentValidateCharacter(
  element: TBoldElement; value: String; representation: Integer;
  Expression: String): Boolean;
begin
  Result := value[1] in ['0'..'9', '-', '+', 'e', 'E', DecimalSeparator];
end;

function Tallform.bsrRentPerResidentValidateString(
  element: TBoldElement; value: String; representation: Integer;
  Expression: String): Boolean;
begin
  try
    StrToCurr(value);
    Result := True;
  except
    Result := False;
  end;
end;

function Tallform.bsrResidentsTotalAssetsGetAsString(
  element: TBoldElement; representation: Integer;
  Expression: String): String;
var
  i: integer;
  sum: Currency;
begin
  Sum := 0;
  if element is TResidential_Building then
    with Element as TResidential_Building do
      for i := 0 to Residents.Count-1 do
        Sum := Sum + Residents[i].Assets;
  Result := CurrToStr(Sum);
end;

procedure Tallform.bsrResidentsTotalAssetsSubscribe(element: TBoldElement;
  representation: Integer; Expression: String;
  Subscriber: TBoldSubscriber);
var
  i: integer;
begin
  if element is TResidential_Building then
    with Element as TResidential_Building do
    begin
      SubscribeToExpression('residents', subscriber, true);
      for i := 0 to Residents.Count-1 do
        Residents[i].SubscribeToExpression('assets', subscriber, false);
    end;
end;

procedure Tallform.bsrAddressSetColor(
  element: TBoldElement; var aColor: TColor; representation: Integer;
  Expression: String);
begin
  if Assigned(element) then
    with element as TBuilding do
    begin
      if Pos('Bold', Address) > 0 then
        aColor := clAqua;
    end;
end;

procedure Tallform.bsrAddressSetFont(element: TBoldElement;
  aFont: TFont; representation: Integer; Expression: String);
begin
  if Assigned(element) then
    with (element as TBuilding).M_Address do
    begin
      if Pos( 'Bold', AsString ) > 0 then
        aFont.Style := aFont.Style + [fsBold];
      if Pos( 'Rose', AsString ) > 0 then
        aFont.Color := clRed;
      if Pos( 'Select', AsString ) > 0 then
        aFont.Color := clGreen;
    end;
end;

procedure Tallform.NewBuilding1Click(Sender: TObject);
begin
  TBuilding.Create(nil);
end;

function Tallform.CurrentBuildingHandle: TBoldListHandle;
begin
  if PageControl1.ActivePage = tabBuilding then
    result := blhAllBuilding
  else
    result := blhAllResidentialBuilding;
end;

procedure Tallform.BoldActivateSystemAction1SystemOpened(Sender: TObject);
begin
  Randomize;
end;

procedure Tallform.HighRentRendererSetColor(Element: TBoldElement;
  var AColor: TColor; Representation: Integer; Expression: String);
begin
  if element is TResidential_Building then
  begin
    if (TResidential_Building(element).totalRent) >= 1500 then
      aColor := clSilver
    else
      aColor := clWhite;
  end;
end;

procedure Tallform.HighRentRendererSetFont(Element: TBoldElement;
  AFont: TFont; Representation: Integer; Expression: String);
begin
  if element is TResidential_Building then
  begin
    if (TResidential_Building(element).totalRent) >= 1500 then
      AFont.Color := clRed
    else
      AFont.Color := clWindowText;
  end;

end;

{function Tallform.BoldPlaceableAFP1GetFormClass(
  Element: TBoldElement): TFormClass;
begin
   result := nil;
  if (Element is TPerson) then
    Result := TPersonAutoForm;
end;

function Tallform.BoldPlaceableAFP1RetrieveHandle(
  Form: TForm): TBoldReferenceHandle;
begin
  Result := nil;
  if (form is TPersonAutoForm) then
    Result := (Form as TPersonAutoForm).brhPerson;
end;}

end.


